src/cargo/ops/cargo_compile.rs

```diff
@@ -1526,19 +1527,38 @@ fn find_named_targets<'a>(
     };
 
     if proposals.is_empty() {
-        let targets = packages.iter().flat_map(|pkg| {
-            pkg.targets()
-                .iter()
-                .filter(|target| is_expected_kind(target))
-        });
-        let suggestion = closest_msg(target_name, targets, |t| t.name());
-        anyhow::bail!(
-            "no {} target {} `{}`{}",
-            target_desc,
-            if is_glob { "matches pattern" } else { "named" },
-            target_name,
-            suggestion
-        );
+        let targets = packages
+            .iter()
+            .flat_map(|pkg| {
+                pkg.targets()
+                    .iter()
+                    .filter(|target| is_expected_kind(target))
+            })
+            .collect::<Vec<_>>();
+        let suggestion = closest_msg(target_name, targets.iter(), |t| t.name());
+        if !suggestion.is_empty() {
+            anyhow::bail!(
+                "no {} target {} `{}`{}",
+                target_desc,
+                if is_glob { "matches pattern" } else { "named" },
+                target_name,
+                suggestion
+            );
+        } else {
+            let mut msg = String::new();
+            writeln!(
+                msg,
+                "no {} target {} `{}`.\n",
+                target_desc,
+                if is_glob { "matches pattern" } else { "named" },
+                target_name,
+            )?;
+            writeln!(msg, "Available {} targets:", target_desc,)?;
+            for target in targets {
+                writeln!(msg, "    {}", target.name())?;
+            }
+            anyhow::bail!(msg);
```

Comment:
> I'll add a unit test later.

*(personally, I find comments on the PR diff easier to track than comments in the PR thread)*

Reply:
Got it. Thanks! I'll add it in PR diff next time.